lib/authenticated-rest-api/authenticated-rest-api.ts

```diff
@@ -0,0 +1,272 @@
+
+import { aws_apigateway as apigateway } from "aws-cdk-lib";
+
+import { aws_cloudwatch as cloudwatch } from "aws-cdk-lib";
+import { aws_cloudwatch_actions as cloudwatchActions } from "aws-cdk-lib";
+import { aws_logs as awslogs } from "aws-cdk-lib";
+import { aws_lambda as lambda } from "aws-cdk-lib";
+import { aws_lambda_nodejs as lambdaNodeJs } from "aws-cdk-lib";
+import { Construct } from "constructs";
+import * as path from "path";
+
+import { AuthenticatedRestApiProps } from "./authenticated-rest-api-props";
+import { ResourceProps } from "./resource-props";
+import { buildLambdaEnvironment } from "../util/build-lambda-environment";
+
+const DEFAULT_API_LATENCY_THRESHOLD = cdk.Duration.minutes(1);
+const DEFAULT_LAMBDA_DURATION_THRESHOLD = cdk.Duration.minutes(1);
+
+export class AuthenticatedRestApi extends Construct {
+  readonly restApiId: string;
+  readonly domainName: apigateway.DomainName;
+
+  private restApi: apigateway.RestApi;
+  private alarmAction: cloudwatch.IAlarmAction;
+
+  constructor(scope: Construct, id: string, props: AuthenticatedRestApiProps) {
+    super(scope, id);
+
+    if (
+      (props.domainName && !props.certificateArn) ||
+      (!props.domainName && props.certificateArn)
+    ) {
+      cdk.Annotations.of(scope).addError(
+        `To use a custom domain name both certificateArn and domainName must be specified`,
+      );
+    }
+    this.domainName = new apigateway.DomainName(this, "domain-name", {
+      domainName: props.domainName,
+      certificate: acm.Certificate.fromCertificateArn(
+        this,
+        "cert",
+        props.certificateArn,
+      ),
+    });
+    const apiName = `${props.prefix}${props.name}`;
+    const apiGatewayProps: apigateway.RestApiProps = {
+      restApiName: apiName,
+      deployOptions: {
+        stageName: props.stageName,
+      },
+      ...(props.corsDomain && {
+        corsPreflight: {
+          allowHeaders: ["*"],
+          allowMethods: ["ANY"],
+          allowCredentials: true,
+          allowOrigins: props.corsDomain,
+        },
+      }),
+    };
+
+    this.restApi = new apigateway.RestApi(this, apiName, apiGatewayProps);
+
+    this.restApiId = this.restApi.restApiId;
+
+    if (props.logging?.enabled) {
+      const logGroup = new awslogs.LogGroup(this, "apiGatewayLogs", {
+        // CloudWatch Logs automatically enables the /aws/vendedlogs/* prefix
+        // in the resource policy, so we use it to avoid adding each log
+        // group to the policy, which has a limit of 5120 characters.
+        logGroupName: `/aws/vendedlogs/${apiName}-accessLog`,
+        retention: props.logging.retention,
+      });
+
+      new cdk.CfnOutput(this, "apiGatewayLogGroup", {
+        exportName: `${apiName}-accessLogGroup`,
+        value: logGroup.logGroupName,
+      });
+    }
+
+    this.alarmAction = new cloudwatchActions.SnsAction(props.alarmTopic);
+
+    // TODO: Validation of scopes is not yet implemented.
```

Comment:
- [ ] Resolve

Reply:
I put in the description of the PR, my intention not to resolve this here. Are you saying this has to be resolved before this PR can be merged?

Reply:
We have a long-standing policy of not leaving TODOs or commented code in our codecases. I'd be fine with you simply removing the TODO.